plutus-chain-index/app/Main.hs

```diff
@@ -9,19 +9,101 @@
+import           Cardano.Protocol.Socket.Client      (ChainSyncEvent (..), runChainSync)
+import           CommandLine                         (AppConfig (..), Command (..), cmdWithHelpParser)
+import           Config                              (ChainIndexConfig)
+import qualified Config                              as Config
+import           Ledger                              (Slot (..))
+import           Logging                             (defaultConfig, loadConfig)
+import           Plutus.ChainIndex.Compatibility     (fromCardanoBlock, fromCardanoTip)
+import           Plutus.ChainIndex.Effects           (ChainIndexControlEffect (..), ChainIndexQueryEffect (..),
+                                                      appendBlock, rollback)
+import           Plutus.ChainIndex.Emulator.Handlers (ChainIndexEmulatorState (..), ChainIndexError (..),
+                                                      ChainIndexLog (..), handleControl, handleQuery)
+import           Plutus.Monitoring.Util              (runLogEffects)
+
+type ChainIndexEffects
+  = '[ ChainIndexControlEffect
+     , ChainIndexQueryEffect
+     , State ChainIndexEmulatorState
+     , Error ChainIndexError
+     , LogMsg ChainIndexLog
+     ]
+
+runChainIndex
+  :: Trace IO ChainIndexLog
+  -> STM.TVar ChainIndexEmulatorState
+  -> Eff ChainIndexEffects a
+  -> IO ()
+runChainIndex trace emulatorState effect = do
+  -- First run the STM block capturing all log messages emited on a
+  -- successful STM transaction.
+  logMessages <- liftIO $ STM.atomically $ do
+    oldEmulatorState <- STM.readTVar emulatorState
+    let (result, logMessages')
+          = interpret handleControl effect
+          & interpret handleQuery
+          & runState oldEmulatorState
+          & runError
+          & reinterpret
+              (handleLogWriter @ChainIndexLog
+                               @(Seq (LogMessage ChainIndexLog)) $ unto pure)
+          & runWriter @(Seq (LogMessage ChainIndexLog))
+          & run
+    case result of
+      Left err ->
+        -- TODO: log the error
+        pure $ LogMessage Error (Err err) <| logMessages'
+      Right (_, newState) -> do
+        STM.writeTVar emulatorState newState
+        pure logMessages'
+  -- Log all previously captured messages
+  traverse_ (send . LMessage) logMessages
+    & runLogEffects trace
+
+chainSyncHandler
+  :: Trace IO ChainIndexLog
+  -> STM.TVar ChainIndexEmulatorState
+  -> ChainSyncEvent
+  -> Slot
+  -> IO ()
+chainSyncHandler trace mState
+  (RollForward block tip) _ = do
+    let ciBlock = fromCardanoBlock block
+    case ciBlock of
+      Left err    ->
+        logError trace (ConversionFailed err)
+      Right block' ->
+        runChainIndex trace mState $ appendBlock (fromCardanoTip tip) block'
+chainSyncHandler trace mState
+  (RollBackward _ tip) _ =
+    -- Do we really want to pass the tip of the new blockchain to the
+    -- rollback function (rather than the point where the chains diverge)?
```

Comment:
I thought `tip` _was_ the point where the chains diverge?

Reply:
There's a bunch of errors that I guess stem from this. I will fix it in the let's test things PR.